File: fsharp/prime-factors/PrimeFactors.fs
module PrimeFactors

    let private cons (head:int) (tail:int list):int list = head::tail

    let rec private factorize (differenceList:int list -> int list) (number:int64) (primeFactor:int):int list =
        match number with
        | 1L -> differenceList []
        | _ ->
            match number % (int64 primeFactor) with
            | 0L -> factorize (differenceList << cons primeFactor) (number / (int64 primeFactor)) primeFactor
            | _ -> factorize differenceList number (primeFactor + 1)

    let factors (number:int64):int list =
        match number > 1L with
        | true ->
            factorize id number 2
        | false -> []

File: solutions/scala/prime-factors/1/src/main/scala/PrimeFactors.scala
object PrimeFactors
{
    private def cons (head:Int) (tail:List[Int]):List[Int] = head::tail

    private def factorize (stack:List[Long], number:Long, primeFactor:Long):List[Long] =
    {
        number match
        {
            case 1L => stack.reverse
            case _ =>
                number % primeFactor match
                {
                    case 0L => factorize(primeFactor::stack, number / primeFactor, primeFactor)
                    case _ => factorize(stack, number,primeFactor + 1)
                }
        }
    }

    def factors (number:Long):List[Long] =
    {
        number > 1L match
        {
            case true => factorize(Nil, number, 2L)
            case false => List.empty
        }
    }
}